bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java

```diff
@@ -803,8 +885,72 @@ void checkAllLedgers() throws BKException, IOException, InterruptedException, Ke
+                        int ackQuorumSize = metadata.getAckQuorumSize();
+                        if (metadata.isClosed()) {
+                            for (Map.Entry<Long, ? extends List<BookieSocketAddress>> ensemble : metadata
+                                    .getAllEnsembles().entrySet()) {
+                                long startEntryIdOfSegment = ensemble.getKey();
+                                List<BookieSocketAddress> ensembleOfSegment = ensemble.getValue();
+                                boolean segmentAdheringToPlacementPolicy = admin.isEnsembleAdheringToPlacementPolicy(
+                                        ensembleOfSegment, writeQuorumSize, ackQuorumSize);
+                                if (!segmentAdheringToPlacementPolicy) {
+                                    metadataCheckEnsembleNotAdheringToPlacementPolicy.inc();
+                                    LOG.warn(
+                                            "For ledger: {}, Segment starting at entry: {}, with ensemble: {} having "
+                                                    + "writeQuorumSize: {} and ackQuorumSize: {} is not adhering to "
+                                                    + "EnsemblePlacementPolicy",
+                                            ledgerId, startEntryIdOfSegment, ensembleOfSegment, writeQuorumSize,
+                                            ackQuorumSize);
+                                }
+                            }
+                        } else {
+                            LOG.debug("Ledger: {} is not yet closed, so skipping the metadata check analysis for now",
+                                    ledgerId);
+                        }
+                        iterCallback.processResult(BKException.Code.OK, null, null);
+                    } else if (BKException
+                            .getExceptionCode(exception) == BKException.Code.NoSuchLedgerExistsException) {
+                        LOG.debug("Ignoring replication of already deleted ledger {}", ledgerId);
+                        iterCallback.processResult(BKException.Code.OK, null, null);
+                    } else {
+                        LOG.warn("Unable to read the ledger: {} information", ledgerId);
+                        iterCallback.processResult(BKException.getExceptionCode(exception), null, null);
+                    }
+                });
+            }
+        };
+        // Reading the result after processing all the ledgers
+        final List<Integer> resultCode = new ArrayList<Integer>(1);
+        ledgerManager.asyncProcessLedgers(ledgerProcessor, new AsyncCallback.VoidCallback() {
+
+            @Override
+            public void processResult(int rc, String s, Object obj) {
+                resultCode.add(rc);
+                metadataCheckLatch.countDown();
+            }
+        }, null, BKException.Code.OK, BKException.Code.ReadException);
+        try {
+            metadataCheckLatch.await();
```

Comment:
Some debugging around this wait will be good; I am worried about indefinite wait scenarios.

Reply:
actually even I'm concerned about this indefinite wait. It is indefinite wait for other checks as well. Have to think about it.

Reply:
We could put an upperbound to the ovetall processing.
The downside is that ledgers with greater id could starve

Reply:
This series of comments seems not addressed

Reply:
@eolivelli @jvrao @sijie i just want to make sure we are on the same page for this.